src/extension/vscode/client.ts

```diff
@@ -0,0 +1,162 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import type { ApolloClient } from "@apollo/client";
+import { getPrivateAccess } from "../../privateAccess";
+import { createActor } from "../actor";
+import { createRpcClient, createRpcHandler } from "../rpc";
+import { getQueries, getMutations } from "../tab/helpers";
+import { loadErrorCodes } from "../tab/loadErrorCodes";
+import type { MessageAdapter } from "../messageAdapters";
+import { handleExplorerRequests } from "../tab/handleExplorerRequests";
+import { setMaxListeners, WeakRef, FinalizationRegistry } from "./polyfills";
+
+type Reason =
+  | "WS_DISCONNECTED"
+  | "CLIENT_GC"
+  | "WS_ERROR"
+  | "UNREGISTER"
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  | (string & {});
+
+export function registerClient(
+  client: ApolloClient<any>,
+  url: string | URL
+): {
+  connected: Promise<void>;
+  unregister: (reason?: Reason) => void;
+  onCleanup: (cleanup: (reason: Reason) => void) => void;
+} {
+  const clientRef = new WeakRef(client);
+  const registered = _registerClient(clientRef, url);
+  const registry = new FinalizationRegistry(registered.unregister);
+  const unregisterToken = {};
+  registry.register(client, "CLIENT_GC", unregisterToken);
+  registered.onCleanup(() => registry.unregister(unregisterToken));
+  return registered;
+}
+
+function makeErrorHandler(cleanup: (reason?: Reason) => void): EventListener {
+  return (e) => {
+    console.error("Apollo DevTools Client encountered WebSocket error:", e);
+    cleanup("WS_ERROR");
+  };
+}
+
+export function _registerClient(
+  clientRef: WeakRef<ApolloClient<any>>,
+  url: string | URL
+): {
+  connected: Promise<void>;
+  unregister: (reason?: Reason) => void;
+  onCleanup: (cleanup: (reason: Reason) => void) => void;
+} {
+  const { signal, cleanup, registerCleanup } = getCleanupController();
+
+  function getClient() {
+    const client = clientRef.deref();
+    if (client) {
+      return getPrivateAccess(client);
+    } else {
+      cleanup("CLIENT_GC");
+    }
+  }
+
+  const ws = new WebSocket(url);
```

Comment:
I'm curious how multiple clients work with separate socket connections. While I think this is fine for outgoing messages (i.e. from the client -> devtools), does creating a separate connection per client mean that the devtools sends messages to every connected socket?

If so, I think we need to be careful that the rpc message handlers aren't all responding to the same message, otherwise whichever handler responds first will be the data displayed in the devtools. That handler may not necessarily be the client that is selected in the dropdown.

The current `hook` script used for the browser devtools is deliberate about using client IDs for this purpose so that we ensure we're always responding with data for the currently selected client.

Can you double check that each of these handlers don't respond with data for multiple clients? This might be a pretty big bug otherwise (at least for thsoe using multiple clients).

Reply:
All the mapping "which client id references which connection" happens on the server - this client would never get messages meant for other clients. But I'll double-check.